c++/src/kj/compat/http.h

```diff
@@ -321,7 +321,38 @@ class HttpHeaders {
     kj::StringPtr statusText;
   };
 
-  kj::Maybe<Request> tryParseRequest(kj::ArrayPtr<char> content);
+  struct ProtocolError {
+    // Represents a protocol error, such as a bad request method or invalid headers. Debugging such
+    // errors is difficult without a copy of the data which we tried to parse, but this data is
+    // sensitive, so we can't just lump it into the error description directly. ProtocolError
+    // provides this sensitive data separate from the error description.
+    //
+    // Currently, ProtocolErrors are only reported when parsing requests from clients, and is only
+    // used by HttpServerErrorHandler::handleClientProtocolError().
+    //
+    // TODO(cleanup): Should maybe not live in HttpHeaders? HttpServerErrorHandler::ProtocolError?
+    //   Or HttpProtocolError? Or maybe we need a more general way of attaching sensitive context to
+    //   kj::Exceptions?
+
+    kj::StringPtr description;
+    // An error description safe for all the world to see.
+
+    kj::ArrayPtr<char> rawContent;
```

Comment:
It looks like this is always exactly the `content` parameter passed into `tryParseRequest()`? Going to all the trouble to have a OneOf with a struct in it so that you can return it back seems unnecessary.

I had imagined that `handleClientProtocolError()` would be given a third parameter which is, like, `kj::ArrayPtr<const byte> buffer`, documented as containing whatever bytes the client sent which the server didn't understand.

Reply:
I'd prefer that approach, too, but it was less obvious how to do that. We'd need to know the header buffer start and end indices in `HttpServer::Connection::loop()` somehow. We could have `HttpInputStreamImpl::readHeader()` save them when it returns, and allow the most-recently-parsed header slice to be retrieved from a function on that class, I suppose?

Reply:
As discussed, will keep the current approach.